File: app/src/main/java/com/openyogaland/denis/pranacoinwallet/SendFragment.java
package com.openyogaland.denis.pranacoinwallet;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

public class SendFragment extends Fragment implements OnClickListener,
                                                      OnSendResponseObtainedListener
{
  // constants
  private final static String BALANCE               = "balance";
  private static final String MY_COMMISSION_ADDRESS = "PBA8J5vGK4G8brcRehTiyxrw9JAHodCj65";
  private static final double TOTAL_COMMISSION_MAX  = 0.1d;
  private static final double API_COMMISSION_AMOUNT = 0.01d;
  
  // fields
  private Context           context;
  private String            idOfUser;
  private EditText          recipientAddressEditText;
  private EditText          sumEditText;
  
  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState)
  {
    View view = inflater.inflate(R.layout.fragment_send, container, false);
    
    // find views by ids
    recipientAddressEditText = view.findViewById(R.id.recipientAddressEditText);
    sumEditText = view.findViewById(R.id.sumEditText);
    Button   scanButton  = view.findViewById(R.id.scanButton);
    Button   sendButton  = view.findViewById(R.id.sendButton);
    
    // set idOfUser
    context = getContext();
    if(context != null)
    {
      idOfUser = PranacoinWallet.getInstance(context).getIdOfUser();
    }
  
    // set listeners
    sendButton.setOnClickListener(this);
    scanButton.setOnClickListener(this);
    
    return view;
  }
  
  @Override
  public void onClick(View view)
  {
    switch (view.getId())
    {
      default:
      case R.id.sendButton:
        if (PranacoinWallet.hasConnection(context))
        {
          // get address and amount to send
          String recipientAddress = recipientAddressEditText.getText().toString();
          String amount           = sumEditText.getText().toString();
          
          if (PranacoinWallet.stringNotEmpty(recipientAddress) &&
              (PranacoinWallet.stringNotEmpty(amount)))
          {
            double balanceAmount           = Double.parseDouble(loadBalance());
            double amountValue             = Double.parseDouble(amount);
            double myCommissionAmountValue = TOTAL_COMMISSION_MAX - (2 * API_COMMISSION_AMOUNT);
  
            // check if user has enough balance for transfer
            if((amountValue + TOTAL_COMMISSION_MAX) > balanceAmount)
            {
              Toast.makeText(context, getString(R.string.not_enough_funds), Toast.LENGTH_SHORT).show();
            }
            else
            {
              // execute users's transfer
              SendSumTask sendSumTask = new SendSumTask(context, idOfUser, recipientAddress, amount);
              sendSumTask.setOnSendResponseObtainedListener(this);
    
              // execute my commission transfer
              String myCommissionAmount = String.valueOf(myCommissionAmountValue);
              new SendSumTask(context, idOfUser, MY_COMMISSION_ADDRESS, myCommissionAmount);
            }
          }
        }
        else
        {
          Toast.makeText(context, R.string.check_internet_connection, Toast.LENGTH_SHORT).show();
        }
        break;
      case R.id.scanButton:
        IntentIntegrator.forSupportFragment(this).initiateScan();
        break;
    }
  }
  
  @Override
  public void onActivityResult(int requestCode, int resultCode, Intent data)
  {
    IntentResult result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
    if(result != null)
    {
      if(result.getContents() == null)
      {
        Toast.makeText(getContext(), R.string.scanning_cancelled, Toast.LENGTH_SHORT).show();
      }
      else
      {
        View view = getView();
        if(view != null)
        {
          recipientAddressEditText = view.findViewById(R.id.recipientAddressEditText);
          recipientAddressEditText.setText(result.getContents());
        }
        Toast.makeText(context, result.getContents(), Toast.LENGTH_SHORT).show();
      }
    }
  }
  
  @Override
  public void onSendResponseObtained(@NonNull String response)
  {
    Toast.makeText(context, response, Toast.LENGTH_SHORT).show();
  }
  
  private String loadBalance()
  {
    Activity activity = getActivity();
    if(activity != null)
    {
      SharedPreferences sharedPreferences = activity.getPreferences(Context.MODE_PRIVATE);
      return sharedPreferences.getString(BALANCE, "");
    }
    return "";
  }
}

File: app/src/main/java/com/openyogaland/denis/pranacoin_wallet_2_0/view/fragment/SendFragment.kt
package com.openyogaland.denis.pranacoin_wallet_2_0.view.fragment

import android.content.Context
import android.content.Context.MODE_PRIVATE
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.zxing.integration.android.IntentIntegrator.forSupportFragment
import com.google.zxing.integration.android.IntentIntegrator.parseActivityResult
import com.google.zxing.integration.android.IntentResult
import com.openyogaland.denis.pranacoin_wallet_2_0.R
import com.openyogaland.denis.pranacoin_wallet_2_0.application.PranacoinWallet2.Companion.hasConnection
import com.openyogaland.denis.pranacoin_wallet_2_0.application.PranacoinWallet2.Companion.log
import com.openyogaland.denis.pranacoin_wallet_2_0.view.dialog.AlertDialogUtil.showAlertDialog
import com.openyogaland.denis.pranacoin_wallet_2_0.viewmodel.MainViewModel
import java.lang.Double.parseDouble

class SendFragment : Fragment() {
    private var recipientAddressEditText: EditText? = null
    private var sumEditText: EditText? = null
    private val mainViewModel: MainViewModel by activityViewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_send, container, false)
        recipientAddressEditText = view.findViewById(R.id.recipientAddressEditText)
        sumEditText = view.findViewById(R.id.sumEditText)
        val scanButton = view.findViewById<Button>(R.id.scanButton)
        val sendButton = view.findViewById<Button>(R.id.sendButton)
        scanButton.setOnClickListener { scanRecipientQRCode() }
        sendButton.setOnClickListener { sendPranacoins() }

        mainViewModel.sendPranacoinsTransactionLiveData.observe(
            viewLifecycleOwner,
            { transactionEventWrapper ->
                transactionEventWrapper
                    .getEventIfNotHandled()
                    ?.let { transaction ->
                        log("SendFragment.onCreateView(): transaction = $transaction")
                        FirebaseCrashlytics.getInstance().setCustomKey(TRANSACTION, transaction)

                        // TODO 0008-5 check ProgressDialog
                        /*activity?.supportFragmentManager?.let { fragmentManager ->
                            hideProgressDialog(fragmentManager)
                        }*/

                        context?.let { context: Context ->
                            showAlertDialog(
                                context,
                                getString(R.string.transfer_status),
                                getString(R.string.transfer_executed_successfully) +
                                        " transaction = $transaction"
                            )
                        }

                        // TODO 0012 show transaction history in HistoryFragment
                    }
            }
        )
        return view
    }

    private fun sendPranacoins() {
        val myCommissionAmountValue = TOTAL_COMMISSION_MAX - 2 * API_COMMISSION_AMOUNT
        val recipientAddress = recipientAddressEditText?.text.toString()
        val amount = sumEditText?.text.toString()
        FirebaseCrashlytics.getInstance().apply {
            setCustomKey(RECIPIENT_ADDRESS, recipientAddress)
            setCustomKey(AMOUNT, amount)
        }
        if (userHasEnoughBalanceForTransfer(recipientAddress, amount)) {
            context?.let { context: Context ->
                // TODO 0009 check internet connectivity
                if (hasConnection(context)) {

                    // TODO 0008-5 check ProgressDialog
                    /*activity?.supportFragmentManager?.let { fragmentManager ->
                        newProgressDialogInstance()
                            .setMessage("Дождитесь подтверждения сервера об отправке пранакоинов")
                            .show(fragmentManager)
                    }*/

                    mainViewModel.sendPranacoins(recipientAddress, amount)
                    mainViewModel.sendPranacoins(
                        MY_COMMISSION_ADDRESS, myCommissionAmountValue.toString()
                    )
                } else { // no internet connection
                    log(
                        "SendFragment.sendPranacoins(): " +
                                getString(R.string.check_internet_connection)
                    )
                    showAlertDialog(
                        requireContext(),
                        getString(R.string.error),
                        getString(R.string.check_internet_connection)
                    )
                }
            }
        } else { // not enough funds or empty field
            log(
                "SendFragment.sendPranacoins(): " +
                        getString(R.string.not_enough_funds_or_empty)
            )
            showAlertDialog(
                requireContext(),
                getString(R.string.error),
                getString(R.string.not_enough_funds_or_empty)
            )
        }
    }

    private fun userHasEnoughBalanceForTransfer(recipientAddress: String, amount: String): Boolean {
        var result = false
        if (recipientAddress.isNotEmpty() && amount.isNotEmpty()) {
            val balance = parseDouble(loadBalance())
            log("SendFragment.onCreateView(): balance = $balance")
            val amountValue = parseDouble(amount)
            if (amountValue + TOTAL_COMMISSION_MAX <= balance) {
                result = true
            }
        }
        return result
    }

    private fun scanRecipientQRCode() = forSupportFragment(this).initiateScan()

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        data: Intent?
    ) {
        parseActivityResult(requestCode, resultCode, data)
            ?.let { result: IntentResult ->
                if (result.contents == null) {
                    log(
                        "SendFragment.onActivityResult(): " +
                                getString(R.string.scanning_cancelled)
                    )
                    showAlertDialog(
                        requireContext(),
                        getString(R.string.error),
                        getString(R.string.scanning_cancelled)
                    )
                }
                result.contents?.let { contents: String ->
                    view?.let { view: View ->
                        recipientAddressEditText = view.findViewById(R.id.recipientAddressEditText)
                        recipientAddressEditText?.setText(contents)
                    }
                    log(
                        "SendFragment.onActivityResult(): " +
                                getString(R.string.scanResult) +
                                contents
                    )
                    showAlertDialog(
                        requireContext(),
                        getString(R.string.success),
                        getString(R.string.scanResult) + contents
                    )
                }
            }
    }

    private fun loadBalance(): String {
        var result = ""
        activity
            ?.getPreferences(MODE_PRIVATE)
            ?.getString(BALANCE, "")
            ?.let { balance: String ->
                result = balance
            }
        return result
    }

    companion object {
        const val BALANCE = "balance"
        const val MY_COMMISSION_ADDRESS = "PBA8J5vGK4G8brcRehTiyxrw9JAHodCj65"
        const val TOTAL_COMMISSION_MAX = 0.1
        const val API_COMMISSION_AMOUNT = 0.001
        private const val RECIPIENT_ADDRESS = "recipient address"
        private const val AMOUNT = "amount"
        private const val TRANSACTION = "transaction"
    }
}